Preserving file permissions (especially executable bits)?

I'm not sure if it's just me or if there is a bug/limitation in this library. Based off the sample code given in library docs, here's what I wrote:

```
do
  bs <- BS.readFile "/tmp/some-executable-binary"
  s <- Zip.mkEntrySelector "some-executable-binary"
  Zip.createArchive "/tmp/zipfile.zip" (Zip.addEntry Zip.Deflate bs s)
```

This creates `/tmp/zipfile.zip`, **but when it is unzipped, `some-executable-binary` no longer has an executable bit set.**

After that, I tried making the following change...

```
  s <- Zip.mkEntrySelector "/tmp/some-executable-binary"
```

**...but it gave the following error:**

```
*** Exception: Cannot build selector from "/tmp/bootstrap"
```

What am I doing wrong?

At the risk of taking this thread off-topic, I tried compiling zip-1.4.0 with lts-12.1 and ran in the following error. Is there some breaking change between the versions of `base` or `ghc` that you might know about?

```
zip               > configure
zip               > Configuring zip-1.4.0...
zip               > build
zip               > Preprocessing library for zip-1.4.0..
zip               > Building library for zip-1.4.0..
zip               > [1 of 5] Compiling Codec.Archive.Zip.CP437
zip               > [2 of 5] Compiling Codec.Archive.Zip.Type
zip               > [3 of 5] Compiling Codec.Archive.Zip.Internal
zip               > [4 of 5] Compiling Codec.Archive.Zip
zip               > [5 of 5] Compiling Codec.Archive.Zip.Unix
zip               > 
zip               > /private/var/folders/j7/fgt692sj0wzg0lvrk86ctnxc0000gn/T/stack77859/zip-1.4.0/Codec/Archive/Zip/Unix.hs:31:24: error:
zip               >     • Couldn't match expected type ‘Word16’ with actual type ‘Word32’
zip               >     • In the second argument of ‘($)’, namely
zip               >         ‘(w `shiftR` 16) .&. 0x0fff’
zip               >       In the expression: CMode $ (w `shiftR` 16) .&. 0x0fff
zip               >       In an equation for ‘toFileMode’:
zip               >           toFileMode w = CMode $ (w `shiftR` 16) .&. 0x0fff
zip               >    |
zip               > 31 | toFileMode w = CMode $ (w `shiftR` 16) .&. 0x0fff
zip               >    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
zip               > 
zip               > /private/var/folders/j7/fgt692sj0wzg0lvrk86ctnxc0000gn/T/stack77859/zip-1.4.0/Codec/Archive/Zip/Unix.hs:42:26: error:
zip               >     • Couldn't match expected type ‘Word32’ with actual type ‘Word16’
zip               >     • In the expression: (0o100000 .|. c) `shiftL` 16
zip               >       In an equation for ‘fromFileMode’:
zip               >           fromFileMode (CMode c) = (0o100000 .|. c) `shiftL` 16
zip               >    |
zip               > 42 | fromFileMode (CMode c) = (0o100000 .|. c) `shiftL` 16
zip               >    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
zip               > 
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+## Zip 1.4.1
+
+* Fixed the build on Mac.
+
 ## Zip 1.4.0
 
 * The “version made by” info inside archive now correctly sets Unix as the

--- a/Codec/Archive/Zip/Unix.hs
+++ b/Codec/Archive/Zip/Unix.hs
@@ -28,7 +28,7 @@ import System.Posix.Types (CMode (..))
 -- @since 1.4.0
 
 toFileMode :: Word32 -> CMode
-toFileMode w = CMode $ (w `shiftR` 16) .&. 0x0fff
+toFileMode attrs = fromIntegral $ (attrs `shiftR` 16) .&. 0x0fff
 
 -- | Convert external attributes to the file info. The function assumes a
 -- regular file and keeps DOS attributes untouched.
@@ -39,4 +39,4 @@ toFileMode w = CMode $ (w `shiftR` 16) .&. 0x0fff
 -- @since 1.4.0
 
 fromFileMode :: CMode -> Word32
-fromFileMode (CMode c) = (0o100000 .|. c) `shiftL` 16
+fromFileMode cmode = (0o100000 .|. fromIntegral cmode) `shiftL` 16

